utils/intel_blit.cpp

```diff
@@ -0,0 +1,564 @@
+  if (info->init) {
+    batch_destroy(info);
+    info->init = 0;
+  }
+  return 0;
+}
+
+int intel_blit_init(struct intel_info *info) {
+  memset(info, 0, sizeof(*info));
+  batch_init(info);
+  info->init = 1;
+  info->mocs.blitter_dst = 2 << 1;
+  info->mocs.blitter_src = 2 << 1;
+  ALOGV("gpubilit init success\n");
+  return 0;
+}
+
+#define ALIGN(value, alignment) ((value + alignment - 1) & ~(alignment - 1))
+
+struct iris_memregion {
+  struct drm_i915_gem_memory_class_instance region;
+  uint64_t size;
+};
+
+struct i915_device {
+  bool initialized;
+  bool has_local_mem;
+  struct iris_memregion vram, sys;
+};
+
+static struct i915_device dev = {
+  .initialized = false,
+};
+
+static inline void
+intel_gem_add_ext(__u64 *ptr, uint32_t ext_name, struct i915_user_extension *ext) {
+  __u64 *iter = ptr;
+  while (*iter != 0) {
+    iter = (__u64 *) &((struct i915_user_extension *)(uintptr_t)*iter)->next_extension;
+  }
+  ext->name = ext_name;
+  *iter = (uintptr_t) ext;
+}
+
+
+static void prelim_i915_bo_update_meminfo(struct i915_device *i915_dev,
+    const struct prelim_drm_i915_query_memory_regions *meminfo) {
+  i915_dev->has_local_mem = false;
+  for (uint32_t i = 0; i < meminfo->num_regions; i++) {
+    const struct prelim_drm_i915_memory_region_info *mem = &meminfo->regions[i];
+    switch (mem->region.memory_class) {
+    case I915_MEMORY_CLASS_SYSTEM:
+      i915_dev->sys.region = mem->region;
+      i915_dev->sys.size = mem->probed_size;
+      break;
+    case I915_MEMORY_CLASS_DEVICE:
+      i915_dev->vram.region = mem->region;
+      i915_dev->vram.size = mem->probed_size;
+      i915_dev->has_local_mem = i915_dev->vram.size > 0;
+      break;
+    default:
+      break;
+    }
+  }
+}
+
+static int intel_update_meminfo(int fd) {
+  if (dev.initialized) {
+    return 0;
+  }
+
+  struct prelim_drm_i915_query_memory_regions *meminfo = nullptr;
+
+  struct drm_i915_query_item item = {
+    .query_id = PRELIM_DRM_I915_QUERY_MEMORY_REGIONS,
+  };
+
+  struct drm_i915_query query = {
+    .num_items = 1,
+    .items_ptr = (uintptr_t)&item,
+  };
+  int ret = drmIoctl(fd, DRM_IOCTL_I915_QUERY, &query);
+  if (ret < 0) {
+    ALOGE("Failed to query PRELIM_DRM_I915_QUERY_MEMORY_REGIONS\n");
+    return -1;
+  }
+  if (item.length <= 0) {
+    return -1;
+  }
+
+  meminfo = static_cast<struct prelim_drm_i915_query_memory_regions *>(calloc(1, item.length));
+  if (!meminfo) {
+    ALOGE("Out of memory\n");
+    return -1;
+  }
+  item.data_ptr = (uintptr_t)meminfo;
+  ret = drmIoctl(fd, DRM_IOCTL_I915_QUERY, &query);
+  if (ret < 0 || item.length <= 0) {
+    free(meminfo);
+    ALOGE("%s:%d DRM_IOCTL_I915_QUERY error\n", __FUNCTION__, __LINE__);
+    return -1;
+  }
+  prelim_i915_bo_update_meminfo(&dev, meminfo);
+  dev.initialized = true;
+  free(meminfo);
+  return 0;
+}
+
+int intel_i915_fd() {
+  static int temp, fd = -1;
+  char device_path[32];
+  if (fd >= 0)
+    return fd;
+  for (int i = 0; i < 64; ++i) {
```

Comment:
intel_i915_fd is called frequently, 64 seems too big, normally we didn't have such many devices